src/v/cluster/scheduling/leader_balancer_constraints.cc

```diff
@@ -0,0 +1,369 @@
+/*
+ * Copyright 2022 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+#include "cluster/scheduling/leader_balancer_constraints.h"
+
+#include "model/metadata.h"
+#include "vassert.h"
+
+namespace cluster::leader_balancer_types {
+
+even_topic_distributon_constraint::even_topic_distributon_constraint(
+  group_id_to_topic_revision_t group_to_topic_rev,
+  shard_index si,
+  const muted_index& mi)
+  : _si(std::move(si))
+  , _mi(mi)
+  , _group_to_topic_rev(std::move(group_to_topic_rev)) {
+    rebuild_indexes();
+    calc_topic_skew();
+    calculate_error();
+}
+
+void even_topic_distributon_constraint::update_index(const reassignment& r) {
+    // _topic_partition_index, _topic_replica_index, and _topic_opt_leaders
+    // will not change from moving leadership. They will only change if the
+    // replicas themselves are moved. Hence no need to update them here.
+
+    auto topic_id = group_to_topic_id().at(r.group);
+
+    // Update _topic_skew and _error
+
+    auto& skew = _topic_skew.at(topic_id);
+    _error -= skew;
+    skew = adjusted_error(skew, topic_id, r.from, r.to);
+    _error += skew;
+
+    // Update _topic_node_index
+
+    auto& groups_from_node = _topic_node_index.at(topic_id).at(r.from.node_id);
+    auto it = std::find_if(
+      groups_from_node.begin(),
+      groups_from_node.end(),
+      [&r](const auto& g_info) { return r.group == g_info.group_id; });
+
+    vassert(it != groups_from_node.end(), "reassigning non-existent group");
+
+    auto moved_group_info = std::move(*it);
+    groups_from_node.erase(it);
+
+    _topic_node_index[topic_id][r.to.node_id].emplace_back(
+      moved_group_info.group_id, r.to, std::move(moved_group_info.replicas));
+
+    // Update _si
+
+    _si.update_index(r);
+}
+
+std::optional<reassignment>
+even_topic_distributon_constraint::recommended_reassignment() {
+    // Sort topics based on topic error here
+    std::vector<decltype(_topic_node_index)::const_iterator> sorted_topics;
```

Comment:
as a matter of style, i'd avoid decltype since this isn't a template. maybe add this some where?

```
using topic_node_index_type =  topic_map<absl::flat_hash_map<model::node_id, std::vector<group_info>>>;
```

Reply:
Great point, though I may be removing the implementation of this function from the PR. It was originally added when the plan was to stick with the heuristic approach to deciding which leader to move and where. However, now that I've choose to randomized hill climbing to minimize the error functions for the constraints this isn't used anymore.